src/models/kinflated_gamma_poisson_model.py

```diff
@@ -0,0 +1,551 @@
+        a randomly chosen user will have an available inventory of n
+        impressions, of which k are shown.
+
+        Args:
+          k:  np.array specifying number of impressions that were seen by the user.
+          p:  Probability that a given impression will be chosen.
+        Returns:
+          For each k, probability that a randomly chosen user will have an inventory
+          of n impressions, of which k are shown.
+        """
+        return np.array(
+            [np.sum([self.knreach(kv, np.arange(kv, len(self._pmf)), p)]) for kv in k]
+        )
+
+    def kplusreach(self, k, p):
+        """Probability that a random user receives k or more impressions.
+
+        Computed as one minus the probability that the user receives k-1 or
+        fewer impressions.
+
+        Args:
+          k:  np.array specifying number of impressions that were seen by the user.
+          p:  Probability that a given impression will be chosen.
+        Returns:
+          For each k, probability that a randomly chosen user will have an inventory
+          of n impressions, of which k are shown.
+        """
+        if k == 0:
+            return 1.0
+        else:
+            return 1.0 - np.sum(self.kreach(np.arange(k), p))
+
+    def expected_value(self):
+        """Expected value of this distribution."""
+        return np.sum(np.arange(len(self._pmf)) * self._pmf)
+
+
+class KInflatedGammaPoissonModel(ReachCurve):
+    """k-inflated Gamma-Poisson reach curve model for underreported counts.
+
+    The k-inflated Gamma-Poisson distribution is like the Gamma-Poisson
+    distribution except that the first k values of the PMF are allowed to
+    be arbitrary.
+
+    It is given by k+3 parameters: alpha, beta, N, a_1, a_2, ..., a_k.
+    We assume that each a_i >= 0, and that a_1 + a_2 + ... + a_k < 1.
+
+    Let GP(n | alpha, beta) be the PMF of the shifted Gamma-Poisson distribution
+    with parameters alpha, beta.  Also, define s = a_1 + a_2 + ... + a_k,
+    and define t = \sum_{i=1}^k GP(i | alpha, beta).  Then, the PMF of the
+    k-inflated Gamma-Poisson distribution is given as
+
+     kGP(n | alpha, beta, a_1, ..., a_k)
+          = a_n if n <= k,
+            (1-s) / (1-t) GP(n | alpha, beta) if n > k.
+
+    As with the Gamma-Poisson distribution, we use the chi2 objective
+    function for assessing goodness of fit.  That is to say, let h[i]
+    be the number of people that were reached i times, and let hbar[i]
+    be the estimated number of people that would be reached i times,
+    given alpha, beta and N.  The objective function that we
+    compute is as follows:
+
+      chi2(h, hbar) = \sum_i (h[i] - hbar[i])^2 / hbar[i]
+
+    One question is what value of k should be chosen.  We propose
+    to consider successive values of k starting from 0, and to stop
+    when a fit is found such that the value of the objective function
+    is smaller than the value of the chi^2 statistic with f_max degrees
```

Comment:
critical value, i.e., <to fill in>% quantile

Reply:
Thanks, updated.